src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java

```diff
@@ -128,19 +133,26 @@ public UnfilteredPartitionIterator search(ReadExecutionController executionContr
         {
             queryContext.addShadowedKeysLoopCount(1L);
             long lastShadowedKeysCount = queryContext.getShadowedPrimaryKeys().size();
-            final var softLimit = controller.currentSoftLimitEstimate();
             ResultRetriever result = queryIndexes.get();
-            UnfilteredPartitionIterator topK = (UnfilteredPartitionIterator)new VectorTopKProcessor(command).filter(result);
-
+            queryContext.resetRowsMatched();
+            UnfilteredPartitionIterator topK = (UnfilteredPartitionIterator) new VectorTopKProcessor(command).filter(result);
             long currentShadowedKeysCount = queryContext.getShadowedPrimaryKeys().size();
             long newShadowedKeysCount = currentShadowedKeysCount - lastShadowedKeysCount;
-            // Stop if no new shadowed keys found
-            // or if we already tried to search beyond the limit for more than the limit + count of new shadowed keys
-            if (newShadowedKeysCount == 0 || exactLimit <= softLimit - newShadowedKeysCount)
+            logger.debug("Shadow loop iteration {}: rows matched: {}, keys shadowed: {}",
+                         queryContext.shadowedKeysLoopCount(),
+                         queryContext.rowsMatched(),
+                         currentShadowedKeysCount);
+            // Stop if no new shadowed keys found or if we already got enough rows
+            if (newShadowedKeysCount == 0 || exactLimit <= queryContext.rowsMatched())
```

Comment:
I just started running tests, and I see that this logic might not be right. The `rowsMatched` represents rows read from storage that were not shadowed. It's not sufficient to know that we've read more than the exactLimit. The previous logic was checking that we read enough rows from each segment.